iml-agent/src/action_plugins/check_stonith.rs

```diff
@@ -8,74 +8,78 @@ use crate::{
 };
 use elementtree::Element;
 use futures::Future;
-use iml_wire_types::{ComponentState, ConfigState, ServiceState, RunState};
+use iml_wire_types::{ComponentState, ConfigState, RunState, ServiceState};
 use std::default::Default;
 use std::process::Command;
 use tracing::{span, Level};
 use tracing_futures::Instrument;
 
+/// This processes <instance_attributes> section of a stonith
+/// "<primitive>" type from the pacemaker cib.
+///
+/// There are two variables of concern in the attributes:
+/// * pcmk_host_check -
+///   - static-list - pcmk_host_list defines comma deliniated list of hosts
+///     controlled by this fencing agent
+///   - dynamic-list (default) - host list is derived from querying fencing
+///     agent (assume same as "none")
+///   - none - any host can fence any host
+/// * pcmk_host_list - list of hosts controlled
+///
+/// Returns: Some("ERROR MESSAGE") on an issue, or None for no issues
+///
 fn inst_attr_problem(elem: &Option<&Element>, node: &String) -> Option<String> {
-    match elem {
-        None => None,
-        Some(instance) => {
-            let mut hc: Option<bool> = None;
-            let mut hl: Option<bool> = None;
-            for e in instance.children() {
-                match e.get_attr("name") {
-                    Some("pcmk_host_list") => match e.get_attr("value") {
-                        None => return None,
-                        Some(val) => {
-                            if val.contains(node) {
-                                match hc {
-                                    Some(true) => return None,
-                                    Some(false) => {
-                                        return Some(format!(
-                                            "{} missing from pcmk_host_list",
-                                            node
-                                        ))
-                                    }
-                                    None => hl = Some(true),
-                                }
-                            } else {
-                                hl = Some(false);
-                            }
-                        }
-                    },
-                    Some("pcmk_host_check") => match e.get_attr("value") {
-                        Some("static-list") => match hl {
-                            Some(true) => return None,
-                            Some(false) => {
-                                return Some(format!("{} missing from pcmk_host_list", node))
-                            }
-                            None => hc = Some(true),
-                        },
-                        _ => return None,
-                    },
-                    _ => (),
+    if let Some(instance) = elem {
```

Comment:
This feels a little awkward to me because an option will either return something or nothing. In this case, the something will always be an error message and None is returned on success. It seems like a `Result` type would fit better here. You could then return Ok(()) when successful or Err with the error message if there is a problem.